### PR TITLE
Default to POST form validation

### PR DESCRIPTION
**Pull request by daniel-beck**
I wonder whether this is worth an escape hatch? I don't think so given how frequent POST requests for these already are. Thoughts?

### Proposed changelog entries

* Switch to sending POST requests by default for form validation URLs

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by jglick — APPROVED** [review]

**Comment by fqueiruga** [comment]
One question, is there any risk to break plugins with this one? For example, if plugins perform a request to their own defined endpoints and depend on the default HTTP method.

**Comment by jglick** [comment]
In principle some plugin could have added a `@GET` annotation to a `doCheckXXX` method and thus be broken by this. I am not sure why anyone would have bothered, though, as this would not have been solving any bug; normally people add `@POST` to defend against CSRF attacks, but there is little reason to add `@GET` except as a matter of documentation.

**Review by sladyn98 — APPROVED** [review]
LGTM :+1:

**Review by fqueiruga — APPROVED** [review]

**Comment by varyvol** [comment]
@daniel-beck this has several approval but it's marked as WiP and on hold? Does it need further work?

**Review by varyvol — APPROVED** [review]

**Review by res0nance — APPROVED** [review]

**Review by timja — APPROVED** [review]

**Review by oleg-nenashev — DISMISSED** [review]
Pull request template should be followed. This security hardening (IIUC) should definitely be present in the changelog and, if there is a risk of regressions, in the upgrade guidelines

**Comment by daniel-beck** [comment]
> security hardening

Nope, as the endpoints would still handle GET requests.

**Review by uhafner — APPROVED** [review]
[combobox.jelly](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/lib/form/combobox.jelly) is missing a `checkMethod` attribute, is this intended?

Otherwise the PR looks good!

**Comment by daniel-beck** [comment]
@uhafner If it's supported internally, it's just a case of missing Jelly doc. Wouldn't be the first time. Will check when I have some time to work on this.

**Comment by daniel-beck** [comment]
@uhafner Thanks for the suggestion. It was in fact just missing Jelly doc, which I added. (Tested via UI Samples Plugin)

**Comment by daniel-beck** [comment]
@oleg-nenashev 

> Pull request template should be followed.

Done, thanks for the explicit reminder. I removed it as this was really just for a PR build, as the comment stated. Now I think I'm feeling better about merging this even without all the pieces in place.

File: core/src/main/resources/lib/form/combobox.jelly
Comment (timja) [inline core/src/main/resources/lib/form/combobox.jelly:62]: according to docs:

>  By default, use="optional" is assumed.

any reason you're adding it to this field only?

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
@daniel-beck if you want this in then can you remove on-hold please?

**Comment by daniel-beck** [comment]
@timja Thanks for the reminder. I don't want this in 2.282, perhaps 2.283 just in case I forgot something.

**Review by jglick — APPROVED** [review]

**Comment by timja** [comment]
@daniel-beck do you want this in soon?

**Comment by daniel-beck** [comment]
@timja Thanks for the reminder. I removed the label.

Thanks for the reviews everyone!

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by miglen** [comment]
> In principle some plugin could have added a `@GET` annotation to a `doCheckXXX` method and thus be broken by this. I am not sure why anyone would have bothered, though, as this would not have been solving any bug; normally people add `@POST` to defend against CSRF attacks, but there is little reason to add `@GET` except as a matter of documentation.

You had actually predicted the future. I personally don't see much value here as those checks should not contain any sensitive information or configurable settings and only be used to validate things.

There is the related issue: https://issues.jenkins.io/browse/JENKINS-65790

**Comment by sabberworm** [comment]
> You had actually predicted the future. I personally don't see much value here as those checks should not contain any sensitive information or configurable settings and only be used to validate things.
> 
> There is the related issue: https://issues.jenkins.io/browse/JENKINS-65790

Maybe the solution isn’t to roll back this change but to look at the status code of the response and handle non-200 responses differently (e.g. by not blindly inserting the response into the DOM as HTML)?

**Comment by daniel-beck** [comment]
> look at the status code of the response and handle non-200 responses differently (e.g. by not blindly inserting the response into the DOM as HTML

I attempted to do that in https://github.com/jenkinsci/jenkins/pull/5333 but based on the bug it isn't effective here.

**Comment by sabberworm** [comment]
Debugging a bit, it seems here’s the part where the styles get added:

https://github.com/jenkinsci/jenkins/blob/2f612fcac9194419e4a73bc2ee5b7ef422b9d6dd/war/src/main/webapp/scripts/hudson-behavior.js#L2929

Interestingly, the code already checks the response status but then decides to parse and add the HTML anyway, just hidden. Which won’t prevent the `style` tags from being evaluated.

**Comment by daniel-beck** [comment]
@sabberworm Nice find, could you submit a PR to fix this?

**Comment by sabberworm** [comment]
> @sabberworm Nice find, could you submit a PR to fix this?

Done, see #5601

**Comment by sabberworm** [comment]
> @sabberworm Nice find, could you submit a PR to fix this?

Done, see #5601

**Pull request merged**
